File: ToolDemo_30/app/src/main/java/com/demon/tool/util/BitmapUtil.java
package com.demon.tool.util;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.drawable.Drawable;
import android.os.Environment;
import android.util.Base64;
import android.view.View;
import android.widget.ScrollView;

import com.google.zxing.BarcodeFormat;
import com.google.zxing.EncodeHintType;
import com.google.zxing.MultiFormatWriter;
import com.google.zxing.WriterException;
import com.google.zxing.common.BitMatrix;
import com.google.zxing.qrcode.QRCodeWriter;

import java.io.BufferedOutputStream;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Hashtable;

public class BitmapUtil {

    /**
     * 压缩图片获取小图
     *
     * @param filePath 图片地址
     * @param width    目标图片宽
     * @param height   目标图片高
     * @param quality  压缩质量
     * @return
     */
    public static Bitmap getSmallBitmap(String filePath, int width, int height, int quality) {
        final BitmapFactory.Options options = new BitmapFactory.Options();
        options.inJustDecodeBounds = true;
        BitmapFactory.decodeFile(filePath, options);

        // Calculate inSampleSize
        options.inSampleSize = calculateInSampleSize(options, width, height);

        // Decode bitmap with inSampleSize set
        options.inJustDecodeBounds = false;

        Bitmap bm = BitmapFactory.decodeFile(filePath, options);
        if (bm == null) {
            return null;
        }
        ByteArrayOutputStream baos = null;
        try {
            baos = new ByteArrayOutputStream();
            bm.compress(Bitmap.CompressFormat.JPEG, quality, baos);

        } finally {
            try {
                if (baos != null)
                    baos.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return bm;
    }

    /**
     * 压缩图片获取小图
     *
     * @param image
     * @param width
     * @param height
     * @return
     */
    public static Bitmap getSmallBitmap(Bitmap image, int width, int height) {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        image.compress(Bitmap.CompressFormat.JPEG, 100, baos);
        // 判断如果图片大于1M,进行压缩避免在生成图片（BitmapFactory.decodeStream）时溢出
        if (baos.toByteArray().length / 1024 > 1024) {
            baos.reset();// 重置baos即清空baos
            image.compress(Bitmap.CompressFormat.JPEG, 50, baos);// 这里压缩50%，把压缩后的数据存放到baos中
        }
        ByteArrayInputStream isBm = new ByteArrayInputStream(baos.toByteArray());
        BitmapFactory.Options newOpts = new BitmapFactory.Options();
        // 开始读入图片，此时把options.inJustDecodeBounds 设回true了
        newOpts.inJustDecodeBounds = true;
        Bitmap bitmap = BitmapFactory.decodeStream(isBm, null, newOpts);
        newOpts.inJustDecodeBounds = false;
        int w = newOpts.outWidth;
        int h = newOpts.outHeight;
        // 缩放比。由于是固定比例缩放，只用高或者宽其中一个数据进行计算即可
        int be = 1;// be=1表示不缩放
        if (w > h && w > width) {// 如果宽度大的话根据宽度固定大小缩放
            be = newOpts.outWidth / width;
        } else if (w < h && h > height) { // 如果高度高的话根据高度固定大小缩放
            be = newOpts.outHeight / height;
        }
        if (be <= 0)
            be = 1;
        newOpts.inSampleSize = be; // 设置缩放比例
        // newOpts.inPreferredConfig = Config.RGB_565;//降低图片从ARGB888到RGB565
        // 重新读入图片，注意此时已经把options.inJustDecodeBounds 设回false了
        isBm = new ByteArrayInputStream(baos.toByteArray());
        bitmap = BitmapFactory.decodeStream(isBm, null, newOpts);
        return bitmap;
    }

    /**
     * 获取小图片大小
     *
     * @param options
     * @param reqWidth
     * @param reqHeight
     * @return
     */
    private static int calculateInSampleSize(BitmapFactory.Options options,
                                             int reqWidth, int reqHeight) {
        // Raw height and width of image
        final int height = options.outHeight;
        final int width = options.outWidth;
        int inSampleSize = 1;

        if (height > reqHeight || width > reqWidth) {

            // Calculate ratios of height and width to requested height and
            // width
            final int heightRatio = Math.round((float) height
                    / (float) reqHeight);
            final int widthRatio = Math.round((float) width / (float) reqWidth);

            // Choose the smallest ratio as inSampleSize value, this will
            // guarantee
            // a final image with both dimensions larger than or equal to the
            // requested height and width.
            inSampleSize = heightRatio < widthRatio ? widthRatio : heightRatio;
        }

        return inSampleSize;
    }

    /**
     * scrollView转为Bitmap ： 可以处理布局超出界面的情况
     *
     * @param scrollView
     * @return
     */
    public static Bitmap getBitmapByView(ScrollView scrollView) {
        int h = 0;
        Bitmap bitmap = null;
        for (int i = 0; i < scrollView.getChildCount(); i++) {
            h += scrollView.getChildAt(i).getHeight();
        }
        bitmap = Bitmap.createBitmap(scrollView.getWidth(), h, Bitmap.Config.RGB_565);
        final Canvas canvas = new Canvas(bitmap);
        canvas.drawColor(Color.WHITE);
        scrollView.draw(canvas);
        return bitmap;
    }

    /**
     * 获取视图的bitmap图片
     *
     * @param v
     * @return
     */
    public static Bitmap getBitmapFromView(View v) {
        Bitmap b = Bitmap.createBitmap(v.getWidth(), v.getHeight(), Bitmap.Config.RGB_565);
        Canvas c = new Canvas(b);
        v.layout(v.getLeft(), v.getTop(), v.getRight(), v.getBottom());
        // Draw background
        Drawable bgDrawable = v.getBackground();
        if (bgDrawable != null)
            bgDrawable.draw(c);
        else
            c.drawColor(Color.WHITE);
        v.draw(c);
        return b;
    }

    /**
     * 通过 Base64 字符串 获取图片资源
     *
     * @param base64
     * @return
     */
    public static Bitmap base64ToBitmap(String base64) {
        byte[] decodeByte = Base64.decode(base64, Base64.DEFAULT);
        Bitmap bitmap = BitmapFactory.decodeByteArray(decodeByte, 0, decodeByte.length);
        return bitmap;
    }

    /**
     * bitmap转为base64
     *
     * @param bitmap
     * @return
     */
    public static String bitmapToBase64(Bitmap bitmap) {
        String result = null;
        ByteArrayOutputStream baos = null;
        try {
            if (bitmap != null) {
                baos = new ByteArrayOutputStream();
                bitmap.compress(Bitmap.CompressFormat.JPEG, 100, baos);

                baos.flush();
                baos.close();

                byte[] bitmapBytes = baos.toByteArray();
                result = Base64.encodeToString(bitmapBytes, Base64.DEFAULT);
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (baos != null) {
                    baos.flush();
                    baos.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return result;
    }

    /**
     * 用于将给定的内容生成成一维码 注：目前生成内容为中文的话将直接报错，要修改底层jar包的内容
     *
     * @param content 将要生成一维码的内容
     * @return 返回生成好的一维码bitmap
     * @throws WriterException WriterException异常
     */
    public static Bitmap createOneDCode(String content, int w, int h) throws WriterException {
        BitMatrix matrix = new MultiFormatWriter().encode(content, BarcodeFormat.CODE_128, w, h);

        int width = matrix.getWidth();
        int height = matrix.getHeight();
        int[] pixels = new int[width * height];

        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {

                if (matrix.get(x, 0)) {
                    pixels[y * width + x] = 0xff000000;
                } else pixels[y * width + x] = 0xffffffff;
            }
        }

        Bitmap bitmap = Bitmap.createBitmap(width, height,
                Bitmap.Config.ARGB_8888);
        bitmap.setPixels(pixels, 0, width, 0, 0, width, height);
        return bitmap;
    }

    /**
     * 字符串转二维码图片
     *
     * @param content
     * @return 二维码图片
     */
    public static Bitmap createQRcodeBitmap(String content) {
        int w = 300;
        int h = 300;
        try {
            //判断URL合法性
            if (StringUtil.isEmpty(content)) {
                return null;
            }
            Hashtable<EncodeHintType, String> hints = new Hashtable<EncodeHintType, String>();
            hints.put(EncodeHintType.CHARACTER_SET, "utf-8");
            //图像数据转换，使用了矩阵转换
            BitMatrix bitMatrix = new QRCodeWriter().encode(content, BarcodeFormat.QR_CODE, w, h, hints);
            int[] pixels = new int[w * h];
            //下面这里按照二维码的算法，逐个生成二维码的图片，
            //两个for循环是图片横列扫描的结果
            for (int y = 0; y < h; y++) {
                for (int x = 0; x < w; x++) {
                    if (bitMatrix.get(x, y)) {
                        pixels[y * w + x] = 0xff000000;
                    } else {
                        pixels[y * w + x] = 0xffffffff;
                    }
                }
            }
            //生成二维码图片的格式，使用ARGB_8888
            Bitmap bitmap = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888);
            bitmap.setPixels(pixels, 0, w, 0, 0, w, h);
            //显示到我们的ImageView上面
//            im1.setImageBitmap(bitmap);
            return bitmap;
        } catch (WriterException e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * 保存字节图片
     *
     * @param context
     * @param imageBytes 字节图片数据
     * @param path       保存的图片路径
     */
    public static void saveImageByte(Context context, byte[] imageBytes, String path) {
        Bitmap temp = BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.length);
        saveBitmap(context, temp, path);
    }

    /**
     * 保存bitmap图片到本地内存
     *
     * @param context
     * @param bmp
     */
    public static void saveBitmap(Context context, Bitmap bmp) {
        String path = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS).getAbsolutePath();
        path = path + File.separator + "test.jpg";
        saveBitmap(context, bmp, path);
    }

    /**
     * 保存bitmap图片到本地内存
     *
     * @param bmp
     * @param path：可以为空，默认到本地DCIM文件加中
     */
    public static void saveBitmap(Context context, Bitmap bmp, String path) {
        File outFile = new File(path);
        BufferedOutputStream outStream = null;
        try {
            outStream = new BufferedOutputStream(new FileOutputStream(outFile));
            bmp.compress(Bitmap.CompressFormat.PNG, 100, outStream);
            outStream.flush();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (outStream != null)
                    outStream.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * 对图片进行旋转处理
     *
     * @param bitmap
     * @param degree
     * @return
     */
    public static Bitmap rotate(Bitmap bitmap, int degree) {
        Matrix matrix = new Matrix();
        matrix.postRotate(degree);
        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, true);
    }

    /**
     * 对图片进行镜像处理
     *
     * @param bitmap
     * @return
     */
    public static Bitmap mirror(Bitmap bitmap) {
        Matrix matrix = new Matrix();
        matrix.postScale(-1f, 1f);
        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, true);
    }
}

File: KotlinTool/app/src/main/java/com/lilanz/kotlintool/utils/BitmapUtil.kt
package com.lilanz.kotlintool.utils

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.view.View
import android.widget.ScrollView
import com.google.zxing.BarcodeFormat
import com.google.zxing.EncodeHintType
import com.google.zxing.MultiFormatWriter
import com.google.zxing.WriterException
import com.google.zxing.qrcode.QRCodeWriter
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.util.*

class BitmapUtil {
    companion object {
        /**
         * scrollView转为Bitmap ： 可以处理布局超出界面的情况
         * @param scrollView
         * @return
         */
        fun getBitmapByView(scrollView: ScrollView): Bitmap? {
            var h = 0
            var bitmap: Bitmap? = null
            for (i in 0 until scrollView.childCount) {
                h += scrollView.getChildAt(i).height
            }
            bitmap = Bitmap.createBitmap(scrollView.width, h, Bitmap.Config.RGB_565)
            val canvas = Canvas(bitmap)
            canvas.drawColor(Color.WHITE)
            scrollView.draw(canvas)
            return bitmap
        }

        /**
         * 获取视图的bitmap图片
         *
         * @param v
         * @return
         */
        fun getBitmapFromView(v: View): Bitmap? {
            val b = Bitmap.createBitmap(v.width, v.height, Bitmap.Config.RGB_565)
            val c = Canvas(b)
            v.layout(v.left, v.top, v.right, v.bottom)
            // Draw background
            val bgDrawable = v.background
            if (bgDrawable != null) bgDrawable.draw(c) else c.drawColor(Color.WHITE)
            v.draw(c)
            return b
        }

        /**
         * 用于将给定的内容生成成一维码 注：目前生成内容为中文的话将直接报错，要修改底层jar包的内容
         *
         * @param content 将要生成一维码的内容
         * @return 返回生成好的一维码bitmap
         * @throws WriterException WriterException异常
         */
        @Throws(WriterException::class)
        fun createOneDCode(content: String?, w: Int, h: Int): Bitmap? {
            val matrix = MultiFormatWriter().encode(content, BarcodeFormat.CODE_128, w, h)
            val width = matrix.width
            val height = matrix.height
            val pixels = IntArray(width * height)
            for (y in 0 until height) {
                for (x in 0 until width) {
                    if (matrix[x, 0]) {
                        pixels[y * width + x] = -0x1000000
                    } else pixels[y * width + x] = -0x1
                }
            }
            val bitmap = Bitmap.createBitmap(
                width, height,
                Bitmap.Config.ARGB_8888
            )
            bitmap.setPixels(pixels, 0, width, 0, 0, width, height)
            return bitmap
        }

        /**
         * 字符串转二维码图片
         *
         * @param content
         * @return 二维码图片
         */
        fun createQRcodeBitmap(content: String?): Bitmap? {
            val w = 300
            val h = 300
            try { //判断URL合法性
                if (StringUtil.isEmpty(content)) {
                    return null
                }
                val hints =
                    Hashtable<EncodeHintType, String?>()
                hints[EncodeHintType.CHARACTER_SET] = "utf-8"
                //图像数据转换，使用了矩阵转换
                val bitMatrix =
                    QRCodeWriter().encode(content, BarcodeFormat.QR_CODE, w, h, hints)
                val pixels = IntArray(w * h)
                //下面这里按照二维码的算法，逐个生成二维码的图片，
//两个for循环是图片横列扫描的结果
                for (y in 0 until h) {
                    for (x in 0 until w) {
                        if (bitMatrix[x, y]) {
                            pixels[y * w + x] = -0x1000000
                        } else {
                            pixels[y * w + x] = -0x1
                        }
                    }
                }
                //生成二维码图片的格式，使用ARGB_8888
                val bitmap = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
                bitmap.setPixels(pixels, 0, w, 0, 0, w, h)
                //显示到我们的ImageView上面
//            im1.setImageBitmap(bitmap);
                return bitmap
            } catch (e: WriterException) {
                e.printStackTrace()
            }
            return null
        }

        /**
         * 保存字节图片
         *
         * @param context
         * @param imageBytes    字节图片数据
         * @param path          保存的图片路径
         */
        fun saveImageByte(context: Context?, imageBytes: ByteArray, path: String?) {
            val temp = BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.size)
            saveBitmap(context, temp, path)
        }

        /**
         * 保存bitmap图片到本地内存
         *
         * @param context
         * @param bmp
         */
        fun saveBitmap(context: Context?, bmp: Bitmap) {
            val path = FileUtil.getPictureFile(context!!)!!.absolutePath
            saveBitmap(context, bmp, path)
        }

        /**
         * 保存bitmap图片到本地内存
         *
         * @param bmp
         * @param path：可以为空，默认到本地DCIM文件加中
         */
        fun saveBitmap(context: Context?, bmp: Bitmap, path: String?) {
            val outFile = File(path)
            var outStream: BufferedOutputStream? = null
            try {
                outStream = BufferedOutputStream(FileOutputStream(outFile))
                bmp.compress(Bitmap.CompressFormat.PNG, 100, outStream)
                outStream.flush()
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                try {
                    outStream?.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
    }
}